src/communicator/communicator.py

```diff
@@ -50,34 +53,78 @@ def __init__(
         ),
     ):
         """Creates a new Communicator object"""
-        Thread.__init__(self)
         self._configuration = sumo_configuration
         self._port = sumo_port
         self._components = components if components is not None else []
         self._max_tick = max_tick
 
-    def run(self):
+    def run(self) -> str:
+        process = self._run.delay(
+            max_tick=self._max_tick,
+            components_pickle=pickle.dumps(self._components),
+            configuration=self._configuration,
+            port=self._port,
+        )
+        return process.id
+
+    @celery.task(bind=True, ignore_result=False)
+    def _run(
+        self: Task,
+        max_tick: int,
+        components_pickle: bytes,
+        configuration: str,
+        port: int,
+    ):
         """Starts sumo (no gui) and connects using traci. The connection has the `default` label."""
 
-        traci.start([checkBinary("sumo"), "-c", self._configuration], port=self._port)
+        components = pickle.loads(components_pickle)
+        traci.start([checkBinary("sumo"), "-c", configuration], port=port)
+
+        sumo_running = True
+        current_tick = 1
+
+        def update_state():
+            self.update_state(
+                state="PROGRESS",
+                meta={
+                    "sumo_running": sumo_running,
+                    "current": current_tick,
+                    "total": max_tick,
+                },
+            )
 
-        self.sumo_running = True
+        update_state()
 
-        while not self._stopped and self._current_tick <= self._max_tick:
-            for component in self._components:
-                component.next_tick(self._current_tick)
-            self._simulation_step()
-            self._current_tick += 1
+        while current_tick <= max_tick:
+            for component in components:
+                component.next_tick(current_tick)
+            traci.simulationStep()
+            current_tick += 1
+            update_state()
 
         traci.close(wait=False)
 
-        self.sumo_running = False
+        sumo_running = False
+        update_state()
 
-    def stop(self):
+    @classmethod
+    def stop(cls, process_id: str):
         """Stopps the simulation requesting a simulation step.
         At most, one simulation step will happen after this request"""
-        self._stopped = True
+        process = AsyncResult(process_id)
+        process.revoke()
```

Comment:
You may need to change the comment regarding the delay of the cancellation

Reply:
Which comment to you mean exactly if I may ask? 🙃

Reply:
Found it, thank you 😇